src/test/ui/self/arbitrary_self_types_lifetime.rs

```diff
@@ -0,0 +1,80 @@
+#![feature(arbitrary_self_types)]
+
+use std::pin::Pin;
+
+#[derive(Debug)]
+struct Foo;
+#[derive(Debug)]
+struct Bar<'a>(&'a Foo);
+
+impl std::ops::Deref for Bar<'_> {
+    type Target = Foo;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Foo {
+    fn a(&self) -> Bar<'_> {
+        Bar(self)
+    }
+
+    fn b(c: &Self) -> Bar<'_> {
+        Bar(c)
+    }
+
+    fn c(self: Bar<'_>) -> Bar<'_> {
+        self
+    }
+
+    fn d(e: Bar<'_>) -> Bar<'_> {
+        e
+    }
+
+    fn e(self: &Self) -> Bar<'_> {
+        Bar(self)
+    }
+
+    fn f(self: Bar<'_>) -> impl std::fmt::Debug {
+        self
+        //~^ ERROR cannot infer an appropriate lifetime
```

Comment:
Consider splitting this case to a different file which can be `// compile-fail` while the others can be `*-pass`.